leap/multistep/__init__.py

```diff
@@ -385,35 +424,176 @@ def rk_bootstrap(self, cb):
 
         from leap.rk import ORDER_TO_RK_METHOD_BUILDER
         rk_method = ORDER_TO_RK_METHOD_BUILDER[self.function_family.order]
-        rk_tableau = tuple(zip(rk_method.c, rk_method.a_explicit))
         rk_coeffs = rk_method.output_coeffs
+        stage_coeff_set_names = ("explicit",)
+        stage_coeff_sets = {"explicit": rk_method.a_explicit}
+        estimate_coeff_set_names = ("main",)
+        estimate_coeff_sets = {"main": rk_coeffs}
+        rhs_funcs = {"explicit": var("<func>"+self.component_id)}
+
+        # Traverse RK stage loop of appropriate order and update state.
+        rk = rk_method(self.component_id, self.state_filter_name)
+        cb = rk.generate_butcher_init(cb, stage_coeff_set_names,
+                                      stage_coeff_sets, rhs_funcs,
+                                      estimate_coeff_set_names,
+                                      estimate_coeff_sets)
+        cb, rhss, est_vars = rk.generate_butcher_primary(cb, stage_coeff_set_names,
+                                                         stage_coeff_sets, rhs_funcs,
+                                                         estimate_coeff_set_names,
+                                                         estimate_coeff_sets)
 
-        # Stage loop (taken from EmbeddedButcherTableauMethodBuilder)
-        rhss = [var("rk_rhs_" + str(i)) for i in range(len(rk_tableau))]
-        for stage_num, (c, coeffs) in enumerate(rk_tableau):
-            if len(coeffs) == 0:
-                assert c == 0
-                cb(rhss[stage_num], rhs_var)
-            else:
-                stage = self.state + sum(self.dt * coeff * rhss[j]
-                                         for (j, coeff)
-                                         in enumerate(coeffs))
+        # Assign the value of the new state.
+        cb(self.state, est_vars[0])
+
+    def determine_bootstrap_length(self):
+
+        # In the explicit case, this is always
+        # equal to history length.
+        bootstrap_length = self.hist_length
+
+        return bootstrap_length
+# }}}
 
-                if self.state_filter is not None:
-                    stage = self.state_filter(stage)
 
-                cb(rhss[stage_num], self.eval_rhs(self.t + c * self.dt, stage))
+# {{{ am method
 
-        # Merge the values of the RHSs.
-        rk_comb = sum(coeff * rhss[j] for j, coeff in enumerate(rk_coeffs))
 
-        state_est = self.state + self.dt * rk_comb
+class AdamsMoultonMethodBuilder(AdamsMethodBuilder):
+    def generate_primary(self, cb):
+
+        from pytools import UniqueNameGenerator
+        name_gen = UniqueNameGenerator()
+        rhs_next_var = var("rhs_next_var")
+        rhs_var_to_unknown = {}
+        unkvar = cb.fresh_var("unk")
+        rhs_var_to_unknown[rhs_next_var] = unkvar
+
+        # In implicit mode, the time history must
+        # include the *next* point in time.
```

Comment:
This doesn't make sense to me. If it includes the future, why is it called "history"? It seems that this was done just so that `emit_adams_integration` could be passed the whole history as it is? If so, I don't agree with that way of writing this code.

Reply:
I did it this way both for that reason and because it does *eventually* become history after this step, but I can think about a different way of doing this.

Reply:
The way of doing this is changed by 4a87ba9, mostly by way of rethinking naming convention to avoid using the term "history" where it's unwarranted, and changing the order in which things are done as a result.